### ISSUE TITLE
cannot replicate NELL-995 MINERVA result [Ongoing]

### ISSUE THREAD

**Issue by gstoica27** [salesforce/MultiHopKG#11]
Hi, apologies for never following up on the original comment, and for adding this issue (again). I'm not sure how to re-open the previous issue #9 .

I have recently been trying again to replicate the results reported with the implementation of MINERVA by the 'point' method but am still having trouble obtaining near reported performance.

I've tried using the 'point' method from NELL-995.sh right out of the box as well as changing its parameters to exactly mirror MINERVA's hyperparameter setup but haven't been able to get close to their reported performances. For instance, the best performance came from NELL-995 right out of the box with (in decimal form):

HITS@1: .544
HITS@3: .735
HITS@5: .775
HITS@10: .811
MRR: .642

To answer the question from a few months ago regarding training ConvE first, I did not, but my understanding is that ConvE/embedding based methods would be needed to train the reward shaping version of MultiHopKG, instead of the 'point' version. Is this correct? Or should I still train ConvE before training 'point' for NELL-995?

Thanks very much for all the help!

**Comment by todpole3** [comment]
@gstoica27 As far as I recall, I got similar results with the vanilla `point` version. Our `point` implementation does not replicate the results in the MINERVA paper -- the MINERVA results reported in our paper were directly copied from their paper.

There are a few differences between our implementations:
1. We are using entity embeddings while MINERVA does not;
2. It is also mentioned here (https://github.com/salesforce/MultiHopKG/issues/9#issuecomment-519342812) that we did false positive filtering differently.

**Comment by gstoica27** [comment]
I see, do you remember by any chance what parameters you used when you ran the 'point' implementation? 

For the differences:
1. Right, but it is still possible to train using only relations right? By adding the 'relation_only' flag in the run arguments
2. I checked that discussion, but I actually believe both your implementation and MINERVA's perform equivalent filtering. It is true that MINERVA labels the entities by PAD, but this just serves as a mask as in your implementation.

**Comment by todpole3** [comment]
@gstoica27 Did you keep a training curve? At which iteration did you observe the `MRR: .642` results?

Also I'm reproducing the results on our side and will get in touch once it's done.

**Comment by todpole3** [comment]
@gstoica27 The configuration file here should be the same we used in the experiment. I reproduced the results locally.

By epoch 2 the results are already much better:
```
Epoch 0
100%|#################################################################################################################| 69/69 [00:46<00:00,  1.48it/s]
Epoch 0: average training loss = -0.358482260037871 entropy = 1.7475357213441063
=> saving checkpoint to '/export/home/xlin/Projects/MultiHopKG/model/NELL-995-point-xavier-n/a-200-200-3-0.001-0.3-0.1-0.3-256-0.05/checkpoint-0.tar'
Epoch 1
100%|#################################################################################################################| 69/69 [00:46<00:00,  1.48it/s]
Epoch 1: average training loss = -0.521502855069497 entropy = 1.3544448403751148
=> saving checkpoint to '/export/home/xlin/Projects/MultiHopKG/model/NELL-995-point-xavier-n/a-200-200-3-0.001-0.3-0.1-0.3-256-0.05/checkpoint-1.tar'
Epoch 2
100%|#################################################################################################################| 69/69 [00:44<00:00,  1.54it/s]
Epoch 2: average training loss = -0.5769490358584067 entropy = 1.2286313596893759
=> saving checkpoint to '/export/home/xlin/Projects/MultiHopKG/model/NELL-995-point-xavier-n/a-200-200-3-0.001-0.3-0.1-0.3-256-0.05/checkpoint-2.tar'
100%|#################################################################################################################| 28/28 [00:13<00:00,  2.16it/s]
Dev set performance: (correct evaluation)
Hits@1 = 0.703
Hits@3 = 0.798
Hits@5 = 0.816
Hits@10 = 0.837
MRR = 0.755
Dev set performance: (include test set labels)
Hits@1 = 0.703
Hits@3 = 0.798
Hits@5 = 0.816
Hits@10 = 0.837
MRR = 0.755
```

**Comment by gstoica27** [comment]
Hi, interesting, I am not getting those results with your implementation. The best I get (with 'relation_only'=True, and other hyperparameters equal to what is in the original nell-995.sh) yields the following plot for MRR. 
![nell_995_mrr](https://user-images.githubusercontent.com/14844582/62992905-25860000-be0a-11e9-8c86-ba7201dc4104.png)

An additional note: The results presented above seem to be consistent however with MINERVA including the entity embedding in network calculation, which as you suggested in a previous comment is not how MINERVA computes NELL-995 (also I believe if you specify 'relation_only' it appends 'ro' to the model saving dir which is not present in the printout shown). Additionally, I could be incorrect with your test, but in my experiments using the original nell-995.sh the dev is similar to as you reported but the test or inference metrics are much lower. Somewhere between 55-60 Hits@1 for instance. 

Were you able by chance to get similar to MINERVA with the 'relation_only' =True? Or when you mentioned that you were able to get results similar to MINERVA is that with the entity and relation embeddings concatenated during network training/are those hyperparameters the same as what is in the nell-995.sh file?

Thanks very much!

**Comment by todpole3** [comment]
> Additionally, I could be incorrect with your test, but in my experiments using the original nell-995.sh the dev is similar to as you reported but the test or inference metrics are much lower.

Ah, I finally see the problem. For NELL-995 getting the correct test result is tricky. The original data release does not have a `dev` set (only `train` and `test`). Hence we randomly sampled a small set of `train` examples as our `dev` and use it for hyperparameter tuning. We then train on a combination of `train` and `dev` and use this model for `test` evaluation.

We made this a note in README. Please follow the steps below to obtain results on `test`.
```
Note for the NELL-995 dataset:

On this dataset we split the original training data into train.triples and dev.triples, and the final model to test has to be trained with these two files combined.

To obtain the correct test set results, you need to add the --test flag to all data pre-processing, training and inference commands.
./experiment.sh configs/nell-995.sh --process_data <gpu-ID> --test
./experiment-emb.sh configs/nell-995-conve.sh --train <gpu-ID> --test
./experiment-rs.sh configs/NELL-995-rs.sh --train <gpu-ID> --test
./experiment-rs.sh configs/NELL-995-rs.sh --inference <gpu-ID> --test
Leave out the --test flag during development.
``` 

> Were you able by chance to get similar to MINERVA with the 'relation_only' =True?

No, our code does not reproduce MINERVA results even with 'relation_only' = True. Besides the masking issue mentioned above, we also modified the final step of standard beam search to return the unique entities when search is done (section 4.3 of our paper). Due to these differences the two implementations are not directly comparable.

**Comment by gstoica27** [comment]
I see, sorry if I misunderstand but from that section in the README I understood that to obtain results on NELL-995 with the 'point' method, we would just need to append the --test flag to preprocessing/training/inference and the metrics I have been reporting here have all been with that flag in place. Based on this, just to double check that my running is correct then, did my result reported in the opening post on inference with '--test' align with your inference results? Thanks a lot for this discussion + help!

Oh I see, I missed the unique beam search component from the paper. Thanks! 

One note aside from that (minor detail) is that I don't think there is any masking problem. It is true that MINERVA labeled their valid target entities by PAD while your code masks them, but both appear to effectively result in the same masking. In MINERVA PAD is treated as a mask in the form: 1 if entity_id = PAD else 0, as in the MultiHopKG code base I believe. This way though the two forms are different, they are actually treated equivalently. I think this is what the other person hinted at too in the last comment.

Thanks for all your help!

**Comment by todpole3** [comment]
> I see, sorry if I misunderstand but from that section in the README I understood that to obtain results on NELL-995 with the 'point' method, we would just need to append the --test flag to preprocessing/training/inference and the metrics I have been reporting here have all been with that flag in place. Based on this, just to double check that my running is correct then, did my result reported in the opening post on inference with '--test' align with your inference results? Thanks a lot for this discussion + help!

With `--test` flag, change num_epochs to 20.
